### PR TITLE
[dv] Add test csr_mem_rw_with_rand_reset

### PR DESCRIPTION
**Pull request by weicaiyang**
Reuse stress_all_rand_reset structure to implement this
Signed-off-by: Weicai Yang <weicai@google.com>

### CODE REVIEW

File: hw/dv/data/tests/csr_tests.hjson [hunk 1]
```diff
@@ -37,6 +37,13 @@
       run_opts: ["+run_same_csr_outstanding"]
       en_run_modes: ["csr_tests_mode"]
     }
+
+    {
+      name: "{name}_csr_mem_rw_with_rand_reset"
+      uvm_test_seq: "{name}_common_vseq"
+      run_opts: ["+run_csr_mem_rw_with_rand_reset"]
+      en_run_modes: ["csr_tests_mode"]
+    }
```
Comment (cindychip) [inline hw/dv/data/tests/csr_tests.hjson:46]: Thanks for doing this Weicai, do you mind adding this to `hw/dv/tools/testplan` so we can directly import it into each IP's testplan?

Comment (weicaiyang) [reply hw/dv/data/tests/csr_tests.hjson:46]: Thanks Cindy, I forgot to add it to testplan again. Added to csr_testplan

File: hw/dv/sv/cip_lib/cip_base_vseq.sv [hunk 2]
```diff
@@ -335,15 +338,18 @@ class cip_base_vseq #(type RAL_T               = dv_base_reg_block,
...
     void'($value$plusargs("stress_seq=%0s", stress_seq_name));
 
     for (int i = 1; i <= num_times; i++) begin
       bit ongoing_reset;
       bit do_read_and_check_all_csrs;
+      `uvm_info(`gfn, $sformatf("running run_stress_all_with_rand_reset_vseq iteration %0d/%0d",
+                                i, num_times), UVM_LOW)
       // use weight arbitration to lower the priority of tl_error seq
-      if (do_tl_err) p_sequencer.tl_sequencer_h.set_arbitration(UVM_SEQ_ARB_WEIGHTED);
+      if (do_tl_err) p_sequencer.tl_sequencer_h.set_arbitration(UVM_SEQ_ARB_STRICT_FIFO);
```
Comment (cindychip) [inline hw/dv/sv/cip_lib/cip_base_vseq.sv:352]: I think this comment need to update since you are not using ARB_WEIGHTED anymore.

Comment (weicaiyang) [reply hw/dv/sv/cip_lib/cip_base_vseq.sv]: Updated comment. Need to have strict order to drive non-blocking TL access for prediction. Thanks.

File: hw/dv/sv/cip_lib/cip_base_vseq.sv [hunk 3]
```diff
@@ -497,6 +506,30 @@ class cip_base_vseq #(type RAL_T               = dv_base_reg_block,
...
+  virtual task run_csr_mem_rw_vseq(int num_times);
+    fork
+      begin
+        csr_excl_item csr_excl = add_and_return_csr_excl("rw");
+        `uvm_info(`gfn, "running csr rw vseq", UVM_HIGH)
+        run_csr_vseq(.csr_test_type("rw"), .csr_excl(csr_excl), .do_rand_wr_and_reset(0));
+      end
+      begin
+        if (cfg.mem_ranges.size > 0) run_mem_partial_access_vseq(num_times);
+      end
```
Comment (sriyerg) [inline hw/dv/sv/cip_lib/cip_base_vseq.sv:520]:
```suggestion
        if (cfg.mem_ranges.size > 0) run_mem_partial_access_vseq(num_times);
```

Comment (weicaiyang) [reply hw/dv/sv/cip_lib/cip_base_vseq.sv:520]: it's in fork ... join, use `begin ... end` for different threads

File: hw/dv/sv/cip_lib/cip_base_vseq.sv [hunk 4]
```diff
@@ -497,6 +506,30 @@ class cip_base_vseq #(type RAL_T               = dv_base_reg_block,
...
+    join
+  endtask
+
+  virtual task run_csr_mem_rw_with_rand_reset_vseq(int num_times);
+    cip_base_vseq #(RAL_T, CFG_T, COV_T, VIRTUAL_SEQUENCER_T) cip_seq;
+    `downcast(cip_seq, this.clone())
```
Comment (sriyerg) [inline hw/dv/sv/cip_lib/cip_base_vseq.sv:526]: I see that you repurposed the `run_stress_all_with_rand_reset_vseq` to run csr_mem_rw with rand reset. I think the current implementation is a little messy and warrants a simplification. Lets discuss this in our next meeting. 

Not sure I understand why is this needed though. Can you not pass `this` directly (`.seq(this) on line 530)?

File: hw/dv/tools/testplans/csr_testplan.hjson [hunk 5]
```diff
@@ -64,7 +64,21 @@
...
       milestone: V1
-      tests: ["{name}{intf}_csr_aliasing",]
+      tests: ["{name}{intf}_csr_aliasing"]
+    }
+    {
+      name: csr_mem_rw_with_rand_reset
+      desc: '''
+            Verify random reset during CSR/memory access.
+            - Run csr_rw sequence to randomly access CSRs
+            - If memory exists, run mem_partial_access in parallel with csr_rw
```
Comment (sriyerg) [inline hw/dv/tools/testplans/csr_testplan.hjson:74]: Does the test actually prove the mem was also reset properly?

**Comment by sriyerg** [comment]
if () begin .. end can be used directly inside fork.

File: hw/dv/sv/cip_lib/cip_base_vseq.sv (hunk 4 above)
Comment (weicaiyang) [reply hw/dv/sv/cip_lib/cip_base_vseq.sv:526]: yes, a little messy to run the sequence of itself, but I didn't find another better way to reuse run_stress_all_with_rand_reset_vseq

at line 527, need to set `common_seq_type`, if not clone a new object, need to set this type to `this`

File: hw/dv/tools/testplans/csr_testplan.hjson (hunk 5 above)
Comment (weicaiyang) [reply hw/dv/tools/testplans/csr_testplan.hjson:74]: I updated cip_base_vseq to use `clone` to preserve mem data from last run, also make sure only save write data when reset isn't active. Now it will verify mem is reset properly.

File: hw/dv/sv/cip_lib/cip_base_vseq.sv (hunk 3 above)
Comment (weicaiyang) [reply hw/dv/sv/cip_lib/cip_base_vseq.sv]: Updated.
